File: day01.cc
// https://adventofcode.com/2021/day/1
#include <bits/stdc++.h>
using namespace std;

template <class T> vector<T> readFileItems(string fname) {
  ifstream fs(fname);
  T tmp;
  vector<T> v;
  while (fs >> tmp)
    v.push_back(tmp);
  return v;
}

template <class T> int countInc(vector<T> vec) {
  if (vec.size() < 2)
    return 0;
  int prev = vec[0];
  int acc = 0;
  for (int i = 1; i < vec.size(); i++) {
    acc += vec[i] > prev;
    prev = vec[i];
  }
  return acc;
}

int part1() {
  const auto inp = readFileItems<int>("day01.txt");
  return countInc(inp);
}

int part2() {
  const auto inp = readFileItems<int>("day01.txt");
  vector<int> wind(inp.size() - 2);
  for (int i = 0; i < inp.size() - 2; i++)
    wind[i] = inp[i] + inp[i + 1] + inp[i + 2];
  return countInc(wind);
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0);
  cout << part1() << endl;
  cout << part2() << endl;
}

File: day01.swift
import Foundation

func getIntArray(_ fname: String) -> [Int] {
    do {
        let lines = try String(contentsOfFile: fname).split(separator: "\n")
        return lines.map({ Int($0) ?? -1 })
    } catch {
        print(error.localizedDescription)
    }
    return []
}

func countInc(_ arr: [Int]) -> Int {
    var count = 0
    var prev = arr[0]
    for v in arr[1...] {
        if prev < v {
            count += 1
        }
        prev = v
    }
    return count
}

func part1() {
    let arr = getIntArray("day01.txt")
    print(countInc(arr))
}

func part2() {
    let arr = getIntArray("day01.txt")
    var wind = Array(repeating: 0, count: arr.count-2)
    for i in 0..<arr.count-2 {
        wind[i] = arr[i] + arr[i+1] + arr[i+2]
    }
    print(countInc(wind))
}

part1()
part2()
